proposals/0054-sessions.md

```diff
@@ -0,0 +1,310 @@
+defaults being `astro-session-id` for the name, and `HttpOnly` and
+`SameSite=Lax` for the flags. The expiry is set to 1 hour by default, but this
+is also configurable.
+
+## Storage
+
+The backend drivers are implemented using
+[Unstorage](https://unstorage.unjs.io/), which provides a storage abstraction
+with drivers for multiple providers.
+
+The session is implemented as a single object stored in the backend, rather than
+individual entries, to avoid the need to make multiple requests to the backend
+for each read or write. This is a tradeoff between performance and consistency,
+but is a common pattern in session handling. The session is fetched in full or
+created when first accessed, and is written back in full at the end of the
+request.
+
+When `session.get()` is called, the session ID is read from the cookie. If there
+is no session ID, a new one is generated. The full session data is then fetched
+as a single value from the backend and deserialized into using devalue. Future
+reads and writes within that request context are done on this local object.
+
+When `session.set()` is called, if there is no session ID, a new one is
+generated. If a local session object has already been fetched, the value is set
+on that. If not, the data is set on a sparse local session object. This is
+merged with the full session if `get()` is called later in the request.
+Otherwise this is done at the end of the request, and the merged object is
+serialized using devalue and written back to the backend.
+
+## Adapter and integration support
+
+Because of the variety of environments to which Astro can be deployed, there is
+no single approach to storage that can be relied upon in all cases. For this
+reason, adapters should provide default session storage drivers where possible.
+Sessions are only available in on-demand rendered pages and API endpoints, so
+there will always be an adapter available. The Node adapter will use filesystem
+storage by default, but this is not suitable for serverless environments. For
+these, the adapter can default to any storage service that is available. For
+example, Netlify may use Netlify Blobs, Vercel may use Vercel KV or Upstash,
+Cloudflare may use Cloudflare Workers KV and Deno Deploy may use Deno KV.
+Integrations can also provide their own storage drivers, and these can be
+auto-configured by the integration.
+
+This is done using the normal integration API, and should be handled in the
+[`astro:config:done` hook](https://docs.astro.build/en/reference/integrations-reference/#astroconfigdone).
+The adapter or integration is responsible for ensuring that they do not
+overwrite any user-defined driver configuration. Adapters may choose to accept
+their own configuration options which they can apply to the storage driver where
+needed. Adapters may provide a storage driver for use in development, or rely on
+the built-in node adapter which is provided by the dev server.
```

Comment:
There can only be one driver, so I don't think there's an issue with conflicts there. Adapters are meant to prefer the user-defined drivers, though I don't think this is something we can enforce when adapters can do what they want to the config.

Reply:
Here's a concrete some example that come to mind:

A  project with a Cloudflare adapter that attempts to use a Netlify blob. What would happen in this occasion? How does Astro behave? It isn't clear *to me* from the RFC:

```js
import cloudflare from "@astrojs/cloudflare"
import {defineConfg} from "astro/config"

defineConfig({
	driver: 'netlify-blobs' // should use nstorage/drivers/netlify-blobs
	adapter: cloudflare()
})
```

In another case, the use of an adapter that **doesn't allow FS storage**:

```js
import cloudflare from "@astrojs/cloudflare" // let's assume `node:fs` isn't compatible with cloudflare
import {defineConfg} from "astro/config"

defineConfig({
	driver: 'fs',
	adapter: cloudflare()
})
```

Reply:
I don't think we need to do checks for people doing things that are clearly not allowed on that platform. Why would somebody choose Netlify blobs if they're not using Netlify? This isn't something we're doing automatically and hiding from them: this is equivalent to somebody choosing the Netlify adapter when they're deploying to Cloudflare, or using `fs` in a page on a site deployed to Cloudflare.

Reply:
> Why would somebody choose Netlify blobs if they're not using Netlify?

Mistakes, junior people who don't know very well what they're learning, possible differences between local behaviour and production behaviour, etc.

I would like to improve that DX if possible. Why can't an adapter accept a predefined list of supported storage?

Reply:
I'd really like to avoid coupling us so closely to the Unstorage driver implementation. One of the design goals for this is to avoid needing to maintain them. If we did that, we'd need to update all adapters whenever they add a new driver, and theoretically audit them all manually whenever that happens.

Reply:
I don't know if its true that you can't use the `netlify-blobs` driver with Cloudflare. If `netlify-blobs` operates over HTTP then more than likely it would work.

There are a few that don't make sense in serverless, like `fs` or the LRU cache (anything in memory).

----

> I would like to improve that DX if possible. Why can't an adapter accept a predefined list of supported storage?

It would make more sense if they provided a list of drivers *not* supported. Like serverless and in-memory. Although that's not even guaranteed to not work either as a serverless function can stay alive an unknown amount of time.

Since you can write your own drivers this wouldn't work anyways; you could write your own in-memory driver and the adapter would have no way of knowing if it works or not.

Reply:
@matthewp 

> I don't know if its true that you can't use the netlify-blobs driver with Cloudflare. If netlify-blobs operates over HTTP then more than likely it would work.

I don't know either, it was just an example to explain what I meant. 

> It would make more sense if they provided a list of drivers not supported

That's a step forward!
___

@ascorbic 

> I'd really like to avoid coupling us so closely to the Unstorage driver implementation. One of the design goals for this is to avoid needing to maintain them. If we did that, we'd need to update all adapters whenever they add a new driver, and theoretically audit them all manually whenever that happens.

I understand that, however we need to strive towards a decent enough DX and UX for our users. Here are some considerations:
- We are responsible for the libraries we use as a service (see `sharp`, `drizzle`, etc.), and we always try to create a certain layer of compatibility for our users.
- Since we don't state which library we rely on (rightfully), Astro is free to change the underneath library (if it ever happens) without changing the user facing APIs. For images, for example, we remove `squoosh`, but the APIs are unchanged - AFAIK.

Reply:
That's a bigger question then. If we want to fully abstract away unstorage to the extent that we don't need to mention it then we need to document every driver ourselves. Now that may be a sensible thing to do, but it will mean we need to decide on a more limited list of ones that we will support.

Reply:
We talked about this on a call, but I'll restate that we can't ensure that a user only uses compatible libraries with Astro. There are a number of places where a user might use a library that doesn't work:

- Remark plugins
- Vite plugins
- Framework components

There's a tradeoff between protecting the user from mistakes and giving them the power to do what they want. I think this is a case where there's a large amount of compatibility already, and users will naturally use drivers that make sense for their host anyways. Adding restrictions would make the API less powerful and unnecessarily prevent legitimate use-cases.

Reply:
`unstorage`s raison d’être is to _be_ the uncoupled interface _and_ provide many implementations. It's used successfully by nitro and I can attest for its ergonomics. I don't think there's a compelling reason to reimplement it just to own the namespace it comes from

Regarding saving devs from themselves, obviously it comes from a good place but it's just a wild goose chase. The higher the guard rails, the more convoluted the hacks people would put in place to skip them. If a junior person who doesn't know very well what they're learning make a mistake then they would learn "the hard way". I personally believe that providing good descriptive errors is much more educational and effective than guardrails. Generally this is something astro is very good at, so it's not an odd expectation.

The suggested sessions API is already 99% existing. Add `unstorage` and use `Astro.locals`, mix in some cookie action and you're golden. The hard part of sessions was the async local context. With an RFC this obvious and implementation this lean it'll be a shame to bikeshed the issue when everyone involved can clearly see the value.


I hope it's ok that I butted in :)